File: qiita_reader_ios/QiitaReader/Data/Repository/Items/Item.swift
//
//  Item.swift
//  QiitaReader
//
//  Created by Shusuke Ota on 2025/2/15.
//

import Foundation

struct Item: Decodable, Identifiable {
    let id: String
    let likesCount: Int
    let tags: [Self.Tag]
    let title: String
    let updatedAtString: String
    let urlString: String
    let user: User

    enum CodingKeys: String, CodingKey {
        case id
        case likesCount = "likes_count"
        case tags
        case title
        case updatedAtString = "updated_at"
        case urlString = "url"
        case user
    }

    var formattedUpdatedAtString: String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ssZ"
        formatter.locale = Locale(identifier: "ja_JP")

        if let date = formatter.date(from: updatedAtString) {
            let outputFormatter = DateFormatter()
            outputFormatter.dateFormat = "yyyy年MM月dd日"
            return outputFormatter.string(from: date)
        }
        return "-"
    }

    struct User: Decodable {
        let id: String
        let profileImageUrlString: String

        enum CodingKeys: String, CodingKey {
            case id
            case profileImageUrlString = "profile_image_url"
        }

        var profileImageUrl: URL? {
            URL(string: profileImageUrlString)
        }

        var name: String {
            "@" + id
        }
    }

    struct Tag: Decodable, Identifiable {
        let id: UUID = UUID()
        let name: String

        enum CodingKeys: String, CodingKey {
            case name
        }
    }
}

File: qiita_reader_android/app/src/main/java/com/shusuke/qiitareader/data/repository/items/Item.kt
package com.shusuke.qiitareader.data.repository.items

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.text.SimpleDateFormat
import java.util.Locale

/**
 * Qiita API v2 の記事アイテム。
 *
 * 参考: https://qiita.com/api/v2/docs#get-apiv2items
 */
@Serializable
data class Item(
    val id: String,
    @SerialName("likes_count") val likesCount: Int,
    val tags: List<Tag>,
    val title: String,
    @SerialName("updated_at") val updatedAtString: String,
    @SerialName("url") val urlString: String,
    val user: User
) {
    val formattedUpdatedAtString: String
        get() = runCatching {
            val input = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssXXX", Locale.JAPAN)
            val output = SimpleDateFormat("yyyy年MM月dd日", Locale.JAPAN)
            output.format(input.parse(updatedAtString)!!)
        }.getOrElse { "-" }

    @Serializable
    data class User(
        val id: String,
        @SerialName("profile_image_url") val profileImageUrlString: String
    ) {
        val name: String get() = "@$id"
    }

    @Serializable
    data class Tag(
        val name: String
    )
}
